solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+    }
+
+    /**
+     * @dev adds a new sample to the accumulator
+     * can only be called by an owner
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     */
+    function addSample(
+        IERC20Token _token,
+        uint256 _n,
+        uint256 _d
+    ) external override validAddress(address(_token)) greaterThanZero(_d) initialized(_token) {
+        require(hasRole(ROLE_OWNER, msg.sender), "ERR_ACCESS_DENIED");
+
+        addSample(_token, Fraction({ n: _n, d: _d }), time());
+    }
+
+    /**
+     * @dev adds a past sample to the accumulator
+     * can only be called by a seeder to add past samples
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     * @param _time the sampling timestamp
+     */
+    function addPastSample(
+        IERC20Token _token,
+        uint256 _n,
+        uint256 _d,
+        uint256 _time
+    ) external override validAddress(address(_token)) greaterThanZero(_d) initialized(_token) {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        require(_time <= time(), "ERR_INVALID_TIME");
+
+        addSample(_token, Fraction({ n: _n, d: _d }), _time);
+    }
+
+    /**
+     * @dev adds past sample to the accumulator
+     * can only be called by a seeder to add past samples
+     *
+     * @param _token the token the data is accumulated for
+     * @param _ns ratio numerators
+     * @param _ds ratio denominators
+     * @param _times sampling timestamps
+     */
+    function addPastSamples(
+        IERC20Token _token,
+        uint256[] calldata _ns,
+        uint256[] calldata _ds,
+        uint256[] calldata _times
+    ) external override validAddress(address(_token)) initialized(_token) {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+
+        uint256 length = _ns.length;
+        require(length == _ds.length && length == _times.length, "ERR_INVALID_LENGTH");
+
+        for (uint256 i = 0; i < length; ++i) {
+            uint256 n = _ns[i];
+            uint256 d = _ds[i];
+            uint256 t = _times[i];
+
+            _greaterThanZero(d);
+            require(t <= time(), "ERR_INVALID_TIME");
+
+            addSample(_token, Fraction({ n: n, d: d }), t);
+        }
+    }
+
+    /**
+     * @dev calculated the TWA from a specific sample time
+     *
+     * @param _token the token the data is accumulated for
+     * @param _startTime the sampling starting time
+     *
+     * @return TWA's numerator and denominator
+     */
+    function timeWeightedAverage(IERC20Token _token, uint256 _startTime)
+        external
+        view
+        override
+        validAddress(address(_token))
+        returns (uint256, uint256)
+    {
+        TokenData storage tokenData = data[_token];
+        require(tokenData.timestamps[_startTime], "ERR_NO_DATA");
+
```

Comment:
No need for `require(_startTime < tokenData.lastSampleTime, "ERR_INVALID_TIME");` here?
If you accept my comment above, then you get this additional statement "for free".

Reply:
Similarly to the above, it's the same "keep all checks outside of the shared private method" approach